File: backend/src/labs/labs.service.ts
import { Injectable, Logger, NotFoundException } from '@nestjs/common';
import { InjectQueue } from '@nestjs/bull';
import type { Queue } from 'bull';
import { PrismaService } from '../prisma/prisma.service';
import { S3Service } from './services/s3.service';
import { LabResultDto } from './dto/lab-result.dto';
import { BiomarkerDto } from './dto/biomarker.dto';

@Injectable()
export class LabsService {
  private readonly logger = new Logger(LabsService.name);

  constructor(
    private readonly prisma: PrismaService,
    private readonly s3Service: S3Service,
    @InjectQueue('ocr-processing') private readonly ocrQueue: Queue,
  ) {}

  async uploadLab(
    file: Express.Multer.File,
    userId: string,
    systemId: string,
    notes?: string,
  ): Promise<LabResultDto> {
    this.logger.log(`Uploading lab for user: ${userId}`);

    const s3Key = this.s3Service.generateKey(systemId, userId, file.originalname);
    const { url } = await this.s3Service.uploadFile(file, s3Key);

    const labResult = await this.prisma.labResult.create({
      data: {
        userId,
        systemId,
        fileName: file.originalname,
        s3Key,
        s3Url: url,
        processingStatus: 'pending',
      },
    });

    await this.ocrQueue.add('extract-text', {
      labResultId: labResult.id,
      s3Key,
    });

    this.logger.log(`Lab uploaded and queued for processing: ${labResult.id}`);

    return this.mapToDto(labResult);
  }

  async getLabResults(userId: string, systemId: string): Promise<LabResultDto[]> {
    const labResults = await this.prisma.labResult.findMany({
      where: {
        userId,
        systemId,
      },
      orderBy: {
        uploadedAt: 'desc',
      },
    });

    return labResults.map((lab) => this.mapToDto(lab));
  }

  async getLabResult(id: string, userId: string, systemId: string): Promise<LabResultDto> {
    const labResult = await this.prisma.labResult.findFirst({
      where: {
        id,
        userId,
        systemId,
      },
    });

    if (!labResult) {
      throw new NotFoundException('Lab result not found');
    }

    return this.mapToDto(labResult);
  }

  async getBiomarkers(
    labResultId: string,
    userId: string,
    systemId: string,
  ): Promise<BiomarkerDto[]> {
    const labResult = await this.prisma.labResult.findFirst({
      where: {
        id: labResultId,
        userId,
        systemId,
      },
    });

    if (!labResult) {
      throw new NotFoundException('Lab result not found');
    }

    const biomarkers = await this.prisma.biomarker.findMany({
      where: {
        labResultId,
      },
      orderBy: {
        testName: 'asc',
      },
    });

    return biomarkers.map((bio) => this.mapBiomarkerToDto(bio));
  }

  async deleteLabResult(
    id: string,
    userId: string,
    systemId: string,
  ): Promise<void> {
    const labResult = await this.prisma.labResult.findFirst({
      where: {
        id,
        userId,
        systemId,
      },
    });

    if (!labResult) {
      throw new NotFoundException('Lab result not found');
    }

    await this.s3Service.deleteFile(labResult.s3Key);

    await this.prisma.labResult.delete({
      where: { id },
    });

    this.logger.log(`Lab result deleted: ${id}`);
  }

  private mapToDto(labResult: any): LabResultDto {
    return {
      id: labResult.id,
      fileName: labResult.fileName,
      uploadedAt: labResult.uploadedAt,
      processingStatus: labResult.processingStatus,
      s3Url: labResult.s3Url,
      rawOcrText: labResult.rawOcrText,
      createdAt: labResult.createdAt,
      updatedAt: labResult.updatedAt,
    };
  }

  private mapBiomarkerToDto(biomarker: any): BiomarkerDto {
    return {
      id: biomarker.id,
      labResultId: biomarker.labResultId,
      testName: biomarker.testName,
      value: biomarker.value,
      unit: biomarker.unit,
      referenceRangeLow: biomarker.referenceRangeLow,
      referenceRangeHigh: biomarker.referenceRangeHigh,
      testDate: biomarker.testDate,
      notes: biomarker.notes,
      createdAt: biomarker.createdAt,
      updatedAt: biomarker.updatedAt,
    };
  }
}

File: app/services/labs_service.py
import logging
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from fastapi import HTTPException, status, UploadFile

from app.models.lab_result import LabResult, Biomarker
from app.schemas.lab import LabResultResponse, BiomarkerResponse
from app.services.s3_service import S3Service
from app.tasks.ocr_tasks import process_ocr_task

logger = logging.getLogger(__name__)


class LabsService:
    def __init__(self, db: AsyncSession):
        self.db = db
        self.s3_service = S3Service()

    async def upload_lab(
        self,
        file: UploadFile,
        user_id: str,
        system_id: str,
        notes: str = None
    ) -> LabResultResponse:
        logger.info(f"Uploading lab for user: {user_id}")

        file_content = await file.read()

        s3_key = self.s3_service.generate_key(system_id, user_id, file.filename)
        result = await self.s3_service.upload_file(file_content, file.filename, s3_key)

        lab_result = LabResult(
            user_id=user_id,
            system_id=system_id,
            file_name=file.filename,
            s3_key=result["key"],
            s3_url=result["url"],
            processing_status="pending",
        )

        self.db.add(lab_result)
        await self.db.commit()
        await self.db.refresh(lab_result)

        process_ocr_task.delay(lab_result.id, s3_key)

        logger.info(f"Lab uploaded and queued for processing: {lab_result.id}")

        return LabResultResponse.from_orm_with_camel(lab_result)

    async def get_lab_results(self, user_id: str, system_id: str) -> list[LabResultResponse]:
        result = await self.db.execute(
            select(LabResult)
            .where(LabResult.user_id == user_id, LabResult.system_id == system_id)
            .order_by(LabResult.uploaded_at.desc())
        )
        lab_results = result.scalars().all()

        return [LabResultResponse.from_orm_with_camel(lab) for lab in lab_results]

    async def get_lab_result(self, lab_id: str, user_id: str, system_id: str) -> LabResultResponse:
        result = await self.db.execute(
            select(LabResult).where(
                LabResult.id == lab_id,
                LabResult.user_id == user_id,
                LabResult.system_id == system_id
            )
        )
        lab_result = result.scalar_one_or_none()

        if not lab_result:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Lab result not found"
            )

        return LabResultResponse.from_orm_with_camel(lab_result)

    async def get_biomarkers(self, lab_id: str, user_id: str, system_id: str) -> list[BiomarkerResponse]:
        result = await self.db.execute(
            select(LabResult).where(
                LabResult.id == lab_id,
                LabResult.user_id == user_id,
                LabResult.system_id == system_id
            )
        )
        lab_result = result.scalar_one_or_none()

        if not lab_result:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Lab result not found"
            )

        result = await self.db.execute(
            select(Biomarker)
            .where(Biomarker.lab_result_id == lab_id)
            .order_by(Biomarker.test_name.asc())
        )
        biomarkers = result.scalars().all()

        return [BiomarkerResponse.from_orm_with_camel(bio) for bio in biomarkers]

    async def delete_lab_result(self, lab_id: str, user_id: str, system_id: str):
        result = await self.db.execute(
            select(LabResult).where(
                LabResult.id == lab_id,
                LabResult.user_id == user_id,
                LabResult.system_id == system_id
            )
        )
        lab_result = result.scalar_one_or_none()

        if not lab_result:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Lab result not found"
            )

        await self.s3_service.delete_file(lab_result.s3_key)

        await self.db.delete(lab_result)
        await self.db.commit()

        logger.info(f"Lab result deleted: {lab_id}")
